File: plugins/android-expert/skills/android-emulator-skill/scripts/log_monitor.py
#!/usr/bin/env python3
"""
Android Log Monitor - ADB Logcat Wrapper

Monitor device logs with filtering.
"""

import argparse
import sys
import subprocess
import signal
from common import resolve_serial, run_adb_command, ADB_PATH

def main():
    parser = argparse.ArgumentParser(description="Monitor Android Logs")
    parser.add_argument("--package", help="Filter by package name (requires app to be running)")
    parser.add_argument("--tag", help="Filter by tag")
    parser.add_argument("--priority", choices=["V", "D", "I", "W", "E", "F"], default="V", help="Minimum priority")
    parser.add_argument("--grep", help="Grep filter")
    parser.add_argument("--clear", "-c", action="store_true", help="Clear logs first")
    parser.add_argument("--serial", "-s", help="Device serial")

    args = parser.parse_args()

    try:
        serial = resolve_serial(args.serial)
    except RuntimeError as e:
        print(f"Error: {e}")
        sys.exit(1)

    if args.clear:
        run_adb_command(["logcat", "-c"], serial)
        print("Logs cleared.")

    cmd = ["logcat", "-v", "color", f"*:{args.priority}"]

    if args.tag:
        cmd = ["logcat", "-v", "color", "-s", args.tag]

    full_cmd = [ADB_PATH]
    if serial:
        full_cmd.extend(["-s", serial])
    full_cmd.extend(cmd)

    if args.package:
        # Get PID of package
        try:
            res = run_adb_command(["shell", "pidof", args.package], serial, check=False)
            pid = res.stdout.strip()
            if pid:
                print(f"Filtering for package {args.package} (PID: {pid})")
                full_cmd.append(f"--pid={pid}")
            else:
                print(f"Package {args.package} not running. Showing all logs.")
        except Exception:
            pass

    grep_pattern = args.grep

    print(f"Running: {' '.join(full_cmd)}")
    try:
        process = subprocess.Popen(full_cmd, stdout=subprocess.PIPE, stderr=sys.stderr, text=True)
        # grep 过滤在 Python 层完成，不依赖 shell 管道
        for line in process.stdout:
            if grep_pattern and grep_pattern not in line:
                continue
            print(line, end="")
        process.wait()
    except KeyboardInterrupt:
        sys.exit(0)

if __name__ == "__main__":
    main()

File: src/components/procedures/sources/android-device-automation/log_monitor.ts
#!/usr/bin/env node
import { defineCliProcedure, procedureEntry } from "../../definition";
import { spawn } from "node:child_process";
import { fileURLToPath } from "node:url";
import { ADB_PATH, resolveSerial, runAdbCommand } from "./common";
import { realpathSync } from "node:fs";

export const procedure = defineCliProcedure({
  id: "android-device-automation-log-monitor",
  entry: procedureEntry(import.meta.url),
  description:
    "流式输出 Android logcat 日志：支持按包名（自动解析 PID）、tag、优先级和 grep 模式过滤，运行直到 Ctrl+C。",
  owners: { skillIds: ["android-device-automation"] },
  target: "scripts/log_monitor.mjs",
  runtime: "node",
  params: [
    {
      flag: "--package",
      type: "字符串",
      description: "按包名过滤（自动解析 PID）",
      required: false,
    },
    {
      flag: "--tag",
      type: "字符串",
      description: "按日志 tag 过滤",
      required: false,
    },
    {
      flag: "--priority",
      type: "V|D|I|W|E|F",
      description: "最低日志优先级",
      required: false,
    },
    {
      flag: "--grep",
      type: "字符串",
      description: "Node 侧行级 grep 过滤",
      required: false,
    },
    {
      flag: "--clear",
      type: "",
      description: "启动前清除日志，传此标志即启用",
      required: false,
    },
    {
      flag: "--serial",
      type: "字符串",
      description: "目标设备序列号",
      required: false,
    },
  ],

  exampleArgs: { args: ["--package", "com.example.app", "--priority", "W"] },
});

export function buildLogcatCommand(
  args: any,
  serial: any,
  pid: any = null,
): any {
  let command: any[] = ["logcat", "-v", "color", `*:${args.priority}`];
  if (args.tag) {
    command = ["logcat", "-v", "color", "-s", args.tag];
  }
  const fullCommand: any[] = [ADB_PATH];
  if (serial) fullCommand.push("-s", serial);
  fullCommand.push(...command);
  if (pid) {
    fullCommand.push(`--pid=${pid}`);
  }
  return fullCommand;
}
function usage(): any {
  return `Monitor Android logs.

Usage: node scripts/log_monitor.mjs [options]

Options:
  --package <name>       Filter by package name when the app is running
  --tag <tag>            Filter by log tag
  --priority <level>     Minimum priority V|D|I|W|E|F (default: V)
  --grep <pattern>       Grep filter applied in Node
  --clear, -c            Clear logs first
  --serial, -s <serial>  Device serial
  --help                 Show this help
`;
}
export function parseArgs(argv: readonly string[]): any {
  const args: Record<string, any> = {
    packageName: null,
    tag: null,
    priority: "V",
    grep: null,
    clear: false,
    serial: null,
    help: false,
  };
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === "--help" || arg === "-h") {
      args.help = true;
      continue;
    }
    if (arg === "--clear" || arg === "-c") {
      args.clear = true;
      continue;
    }
    if (
      ["--package", "--tag", "--priority", "--grep", "--serial", "-s"].includes(
        arg,
      )
    ) {
      const value = argv[index + 1];
      if (value == null || value.startsWith("--"))
        throw new Error(`${arg} requires a value`);
      index += 1;
      if (arg === "--package") args.packageName = value;
      if (arg === "--tag") args.tag = value;
      if (arg === "--priority") args.priority = value;
      if (arg === "--grep") args.grep = value;
      if (arg === "--serial" || arg === "-s") args.serial = value;
      continue;
    }
    throw new Error(`unrecognized argument: ${arg}`);
  }
  if (!["V", "D", "I", "W", "E", "F"].includes(args.priority)) {
    throw new Error("--priority must be one of: V, D, I, W, E, F");
  }
  return args;
}
export function resolvePackagePid(packageName: any, serial: any): any {
  const result = runAdbCommand(["shell", "pidof", packageName], serial, {
    check: false,
  });
  return (result.stdout ?? "").trim();
}
export function main(argv: readonly string[]): any {
  const args = parseArgs(argv);
  if (args.help) {
    console.log(usage());
    return 0;
  }
  const serial = resolveSerial(args.serial);
  if (args.clear) {
    runAdbCommand(["logcat", "-c"], serial);
    console.log("Logs cleared.");
  }
  let pid: any = null;
  if (args.packageName) {
    try {
      pid = resolvePackagePid(args.packageName, serial);
      if (pid) {
        console.log(`Filtering for package ${args.packageName} (PID: ${pid})`);
      } else {
        console.log(
          `Package ${args.packageName} not running. Showing all logs.`,
        );
      }
    } catch {
      // Keep old behavior: package PID lookup failure should not stop logcat.
    }
  }
  const command = buildLogcatCommand(args, serial, pid);
  console.log(`Running: ${command.join(" ")}`);
  const child = spawn(command[0], command.slice(1), {
    stdio: ["ignore", "pipe", "inherit"],
  });
  child.stdout.setEncoding("utf-8");
  child.stdout.on("data", (chunk: any) => {
    for (const line of chunk.split(/(?<=\n)/)) {
      if (!line) continue;
      if (args.grep && !line.includes(args.grep)) continue;
      process.stdout.write(line);
    }
  });
  child.on("exit", (code: any) => {
    process.exitCode = code ?? 0;
  });
  child.on("error", (error: any) => {
    console.error(`Error: ${error.message}`);
    process.exitCode = 1;
  });
  process.on("SIGINT", () => {
    child.kill("SIGINT");
    process.exitCode = 0;
  });
  return 0;
}
